p2p/transport/webrtc/datachannel.go

```diff
@@ -0,0 +1,402 @@
+package libp2pwebrtc
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"io"
+	"net"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-msgio/protoio"
+	"github.com/pion/datachannel"
+	"github.com/pion/webrtc/v3"
+
+	pb "github.com/libp2p/go-libp2p/p2p/transport/webrtc/pb"
+)
+
+var _ network.MuxedStream = &dataChannel{}
+
+const (
+	// maxMessageSize is limited to 16384 bytes in the SDP.
+	maxMessageSize int = 16384
+	// Pion SCTP association has an internal receive buffer of 1MB (roughly, 1MB per connection).
+	// We can change this value in the SettingEngine before creating the peerconnection.
+	// https://github.com/pion/webrtc/blob/v3.1.49/sctptransport.go#L341
+	maxBufferedAmount int = 2 * maxMessageSize
+	// bufferedAmountLowThreshold and maxBufferedAmount are bound
+	// to a stream but congestion control is done on the whole
+	// SCTP association. This means that a single stream can monopolize
+	// the complete congestion control window (cwnd) if it does not
+	// read stream data and it's remote continues to send. We can
+	// add messages to the send buffer once there is space for 1 full
+	// sized message.
+	bufferedAmountLowThreshold uint64 = uint64(maxBufferedAmount) / 2
+
+	protoOverhead  int = 5
+	varintOverhead int = 2
```

Comment:
What are these? Are we hardcoding assumptions about the size of a resulting varint here? We should avoid that.

If that's not possible, at the very least we need comments explaining what's going on here.